build: move parallelism to the top level Makefile

jordanlewis opened the pull request:
Previously, each submake was executed with -j $(NCPUS). This was slow -
possibly because booting a parallel make with a jobserver is slow.

Instead, we give the top-level make a default parallelism of $(NCPUS).
Thanks to Make's jobserver, the parallelism will be shared properly
among all submake invocations. NCPUS can be overriden on the
commandline in case the parallel output makes it difficult to debug a
problem.

This improves the time it takes `make test` to run all of the parts before `go test` gets
executed (submakes in c-deps) from ~4 seconds to ~1 second on my machine.

jordanlewis committed:
build: move parallelism to the top level Makefile

Previously, each submake was executed with -j $(NCPUS). This was slow -
possibly because booting a parallel make with a jobserver is slow.

Instead, we give the top-level make a default parallelism of $(NCPUS).
Thanks to Make's jobserver, the parallelism will be shared properly
among all submake invocations. NCPUS can be overriden on the
commandline in case the parallel output makes it difficult to debug a
problem.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17098)
<!-- Reviewable:end -->

benesch commented:
/cc @tamird who did this work to begin with

benesch commented:
Approach LGTM, but lemme take a quick peek at how this affects the output of a fresh build.

benesch commented:
And thank you, @jordanlewis, for looking into this!

benesch commented:
Ok, a couple thoughts. On my machine, a noop `make` isn't nearly as slow as on yours. With the following patch applied

```diff
diff --git a/Makefile b/Makefile
index c7d91220cf..90799ca324 100644
--- a/Makefile
+++ b/Makefile
@@ -174,7 +174,7 @@ $(COCKROACH) build buildoss go-install: override LINKFLAGS += \
 build: ## Build the CockroachDB binary.
 buildoss: ## Build the CockroachDB binary without any CCL-licensed code.
 $(COCKROACH) build buildoss go-install: $(CGO_FLAGS_FILES) $(BOOTSTRAP_TARGET) .buildinfo/tag .buildinfo/rev
-	 $(XGO) $(BUILDMODE) -v $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LINKFLAGS)' $(BUILDTARGET)
+	true
 
 .PHONY: install
 install: ## Install the CockroachDB binary.
```

it only takes me about 1s:

```shell
$ time make
GOPATH set to /Users/benesch/go
GEN pkg/cli/zcgo_flags.go
GEN pkg/server/status/zcgo_flags.go
GEN pkg/storage/engine/zcgo_flags.go
GEN pkg/ccl/storageccl/engineccl/zcgo_flags.go
make[1]: Nothing to be done for `build_lib_static'.
[100%] Built target libprotobuf
[ 62%] Built target snappy
[100%] Built target snappy-unittest
[  0%] Built target build_version
[100%] Built target rocksdb
[100%] Built target roach
[ 84%] Built target roach
[100%] Built target roachccl
true

real	0m1.278s
user	0m0.755s
sys	0m0.397s
```

With your patch applied, I see only a slight speedup:

```make
$ time make
GOPATH set to /Users/benesch/go
GEN pkg/cli/zcgo_flags.go
GEN pkg/server/status/zcgo_flags.go
GEN pkg/storage/engine/zcgo_flags.go
GEN pkg/ccl/storageccl/engineccl/zcgo_flags.go
make[1]: Nothing to be done for `build_lib_static'.
[ 62%] Built target snappy
[100%] Built target snappy-unittest
[ 84%] Built target roach
Built target roach
Built target roachccl
[100%] Built target libprotobuf
[  0%] Built target build_version
[100%] Built target rocksdb
:

real	0m0.759s
user	0m0.805s
sys	0m0.428s
```

This is approximately the same speedup as if I handle all `-j` flags myself:

```
$ time make -j8
GOPATH set to /Users/benesch/go
GEN pkg/cli/zcgo_flags.go
GEN pkg/server/status/zcgo_flags.go
GEN pkg/storage/engine/zcgo_flags.go
GEN pkg/ccl/storageccl/engineccl/zcgo_flags.go
make[1]: Nothing to be done for `build_lib_static'.
[ 62%] Built target snappy
[100%] Built target snappy-unittest
[ 84%] Built target roach
Built target roach
Built target roachccl
[100%] Built target libprotobuf
[  0%] Built target build_version
[100%] Built target rocksdb
:

real	0m0.732s
user	0m0.792s
sys	0m0.420s
```

Also, FWIW, this patch doesn't seem to work with the latest version of Make:

```
$ make --version
GNU Make 3.81
$ gmake --version
GNU Make 4.2.1
$ time gmake
GOPATH set to /Users/benesch/go
gmake[1]: Nothing to be done for 'build_lib_static'.
[100%] Built target libprotobuf
[ 62%] Built target snappy
[100%] Built target snappy-unittest
[  0%] Built target build_version
[100%] Built target rocksdb
[100%] Built target roach
[ 84%] Built target roach
[100%] Built target roachccl
GEN pkg/cli/zcgo_flags.go
GEN pkg/server/status/zcgo_flags.go
GEN pkg/storage/engine/zcgo_flags.go
GEN pkg/ccl/storageccl/engineccl/zcgo_flags.go

real	0m1.216s
user	0m0.687s
sys	0m0.396s
```

So I guess my first question is: what version of Make are you running, and why is it so slow?

jordanlewis commented:
I tried with stock osx `make` and latest homebrew `gmake` and had similar slowness. Perhaps it's a problem with my work machine? I can't reproduce this on my home machine, actually. So that's something I need to look into.

On the other hand, I still think it's better to have the top-level make be a jobserver for the submakes so they can all share a pool of parallelism. That's how make is designed to work, after all.

benesch commented:
Yeah, I totally hear you, and I hope I'm not coming across as discouraging. I just worry about parallelism hiding failures, because Make will let other jobs keep printing output even when one has failed. E.g., https://gist.github.com/benesch/de15d6adcfe1bf8a1c80133cab636f42.

knz commented:
@benesch this patch accelerate the build of the C deps for me by at least 30% because now their builds can overlap. I would buy this change just for that benefit.

jordanlewis committed:
build: add missing dependency arrow

libroachccl needs to depend on libroach in the top level makefile, or
else when running in parallel those two targets will both try to build
libroach at once.

benesch commented:
My [slight] preference is still that anyone who wants to deal with the headache of parallel top-level make should set

```shell
export MAKEFLAGS=-jNCPUS
```

in their shell startup script. In a perfect world we wouldn't have ever passed -jNCPUS for the submakes, either.

Anyway, @jordanlewis has convinced me this is worth merging. If it becomes a real headache, we can always revert.

knz commented:
Your suggestion to set MAKEFLAGS in the environment doesn't seem wise to me. If you do this, then the proper sharing of the parallel build scheduler between a top-level make and sub-makes would break. -j should be passed either via a shell *alias* (e.g. `alias make='make -jN'`) or explicitly on the command line.

knz commented on Makefile:
```diff
@@ -46,6 +46,9 @@ SUFFIX       :=
 bindir       := $(prefix)/bin
+NCPUS        = $(shell $(LOCAL_BIN)/ncpus)
+
+MAKEFLAGS += --jobs=$(NCPUS)
```
What about:
```make
PARALLEL = -j$(NCPUS)
MAKEFLAGS += $(PARALLEL)
```

So that the user can override on the command line with `make PARALLEL=`

jordanlewis commented:
Okay, I deleted the modification of `MAKEFLAGS` because it didn't consistently work across versions of `make` and relied on `ncpus` which isn't always available when it's required.

This permits usage of `-j` by users and CI.

benesch commented:
I'm very 👍 on this! Two thoughts, though:

  * Can you also remove `pkg/cmd/ncpus`? AFAIK it's not used by anything else.
  * I think we absolutely need to print a warning if `-j` is not detected, as `make build` is about to get very very slow by default.

jordanlewis commented:
PTAL, now properly detects ncpus without the external binary and prints a warning if the detection fails. Also manually specifying -j will override the detection.

Removed pkg/cmd/ncpus.

tamird commented:
Reviewed 1 of 3 files at r1, 4 of 4 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[build/jflag.sh, line 1 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17098#-KpQxG1UHwyOVJXoC5os:-KpQwg2bGH76JkmOLzHu:b7bvh0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2386ccdf9cdf21d00d369f3456018c890660902/build/jflag.sh#L1)):*
> ```Shell
> #!/usr/bin/env bash
> ```

why none of the usual bash strictness?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17098)*
<!-- Sent from Reviewable.io -->

benesch reviewed (approved):
LGTM!

I think this is worth doing even given our try-not-to-change-the-build-system attitude, since the patch is so small, but would love for someone else to sign off as well.

benesch commented:
@tamird means `set -euo pipefail`

tamird commented:
Why would it exit? `false || true` doesn't exit in a script that specifies `-e`, IIRC.

benesch commented:
-e should be ok, because the final echo in the `||` chain will always
succeed.

On Wed, Jul 19, 2017 at 1:59 PM, Jordan Lewis <notifications@github.com>
wrote:

> Can't use -e because then the script would exit when trying to detect the
> number of cpus. But no reason why we couldn't use -uo pipefail I suppose.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17098#issuecomment-316467775>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IKZXfRCZn-an7vsDDxN6v9aFaoG5ks5sPkQegaJpZM4OcFi4>
> .
>

jordanlewis commented:
Y'all right. Done.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17098#-:-KpR1WmekNKqKQMdRhTV:bnfp4nl)*
<!-- Sent from Reviewable.io -->
